test/common/io/io_uring_impl_test.cc

```diff
@@ -0,0 +1,239 @@
+#include <sys/epoll.h>
+
+#include "source/common/io/io_uring_impl.h"
+
+#include "test/test_common/environment.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Io {
+namespace {
+
+class IoUringBaseTest : public ::testing::Test {
+public:
+  void TearDown() override {
+    auto& uring = factory_.getOrCreateUring();
+    if (uring.isEventfdRegistered()) {
+      uring.unregisterEventfd();
+    }
+  }
+
+  static thread_local const IoUringFactoryImpl factory_;
+};
+
+thread_local const IoUringFactoryImpl IoUringBaseTest::factory_(2, false);
+
+class IoUringImplParamTest
+    : public IoUringBaseTest,
+      public testing::WithParamInterface<std::function<IoUringResult(IoUring&, os_fd_t)>> {};
+
+INSTANTIATE_TEST_SUITE_P(InvalidPrepareMethodParamsTest, IoUringImplParamTest,
+                         testing::Values(
+                             [](IoUring& uring, os_fd_t fd) -> IoUringResult {
+                               return uring.prepareAccept(fd, nullptr, nullptr, nullptr);
+                             },
+                             [](IoUring& uring, os_fd_t fd) -> IoUringResult {
+                               auto address =
+                                   std::make_shared<Network::Address::EnvoyInternalInstance>(
+                                       "test");
+                               return uring.prepareConnect(fd, address, nullptr);
+                             },
+                             [](IoUring& uring, os_fd_t fd) -> IoUringResult {
+                               return uring.prepareReadv(fd, nullptr, 0, 0, nullptr);
+                             },
+                             [](IoUring& uring, os_fd_t fd) -> IoUringResult {
+                               return uring.prepareWritev(fd, nullptr, 0, 0, nullptr);
+                             },
+                             [](IoUring& uring, os_fd_t fd) -> IoUringResult {
+                               return uring.prepareClose(fd, nullptr);
+                             }));
+
+TEST_P(IoUringImplParamTest, InvalidParams) {
+  os_fd_t fd;
+  SET_SOCKET_INVALID(fd);
+
+  auto& uring = factory_.getOrCreateUring();
+
+  os_fd_t event_fd = uring.registerEventfd();
+  os_fd_t epoll_fd = epoll_create1(0);
+  ASSERT_FALSE(epoll_fd == -1);
+  struct epoll_event ev, events[10];
+  ev.events = EPOLLIN | EPOLLET;
+  ev.data.fd = event_fd;
+  ASSERT_FALSE(epoll_ctl(epoll_fd, EPOLL_CTL_ADD, event_fd, &ev) == -1);
+
+  auto prepare_method = GetParam();
+  IoUringResult res = prepare_method(uring, fd);
+  EXPECT_EQ(res, IoUringResult::Ok);
+  res = prepare_method(uring, fd);
+  EXPECT_EQ(res, IoUringResult::Ok);
+  res = prepare_method(uring, fd);
+  EXPECT_EQ(res, IoUringResult::Failed);
+  res = uring.submit();
+  EXPECT_EQ(res, IoUringResult::Ok);
+  res = uring.submit();
+  EXPECT_EQ(res, IoUringResult::Ok);
+
+  int ret = epoll_wait(epoll_fd, events, 10, -1);
```

Comment:
Use of epoll_wait to decide when to drive the IoUring seems like generic code that most uring users will needed.  Should this use of epoll_wait be moved to IoUringImpl ?

Reply:
Users don't need epoll_wait if they rely on libevent for event handling (via Dispatcher).

Reply:
If the intended use is together with the dispatcher, perhaps this test should use the dispatcher for event handling.

Reply:
Good point. The test doesn't include `sys/epoll.h` now, event handlers are run by dispatcher instances.